matrix_content_scanner/scanner/file_downloader.py

```diff
@@ -0,0 +1,282 @@
+        endpoint_version: str = "v3",
+        thumbnail_params: Optional[Dict[str, List[str]]] = None,
+    ) -> str:
+        """Turn a `server_name/media_id` path into an https:// one we can use to fetch
+        the media.
+
+        Note that if `base_homeserver_url` is set to an http URL, it will not be turned
+        into an https one.
+
+        Args:
+            media_path: The media path to translate.
+            endpoint_version: The version of the download endpoint to use. As of Matrix
+                v1.1, this is either "v3" or "r0".
+            thumbnail_params: If present, then we want to request and scan a thumbnail
+                generated with the provided parameters instead of the full media.
+
+        Returns:
+            An https URL to use. If `base_homeserver_url` is set in the config, this
+            will be used as the base of the URL.
+        """
+        server_name, media_id = media_path.split("/")
+
+        # Figure out what base URL to use. If one is specified in the configuration file,
+        # use it, otherwise try to discover one using .well-known. If that fails, use the
+        # server name with an HTTPS scheme.
+        if self._base_url is not None:
+            base_url = self._base_url
+        else:
+            base_url = "https://" + server_name
+
+        prefix = self.MEDIA_DOWNLOAD_PREFIX
+        query = None
+        if thumbnail_params is not None:
+            # If there are thumbnail generation parameters, then we want to generate a
+            # thumbnail of the file, not download its full-sized version.
+            prefix = self.MEDIA_THUMBNAIL_PREFIX
+
+            # Reconstruct the query parameters provided by the client.
+            query = ""
+            for key, items in thumbnail_params.items():
+                for item in items:
+                    query += "%s=%s&" % (
+                        urllib.parse.quote(key),
+                        urllib.parse.quote(item),
+                    )
+            query = query[:-1]
+
+        # Build the full URL.
+        path_prefix = prefix % endpoint_version
+        url = "%s/%s/%s/%s" % (
+            base_url,
+            path_prefix,
+            urllib.parse.quote(server_name),
+            urllib.parse.quote(media_id),
+        )
+        if query is not None:
+            # If there are query parameters, append them to the URL.
+            url += "?%s" % query
+
+        return url
+
+    async def _get_file_content(self, url: str) -> MediaDescription:
+        """Retrieve the content of the file at a given URL.
+
+        Args:
+            url: The URL to query.
+
+        Returns:
+            A description of the file (including its full content).
+
+        Raises:
+            _PathNotFoundException: the server returned an error that can mean the path
+                of the request wasn't understood, e.g. because we requested a v3 URL but
+                the server only supports r0.
```

Comment:
can't it also mean that the media does not exist?

Reply:
Yes, I've enriched the comment to explain why we need a separate error class for this.